core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/MemorySailStore.java

```diff
@@ -746,4 +745,19 @@ private boolean isContextResource(MemResource memResource, int snapshot) throws
 			}
 		}
 	}
+
+	/**
+	 * increments the current snapshot by 1 and schedules a snapshot cleanup if necessary.
+	 */
+	protected void incrementSnapshot() {
+		txnLockManager.lock();
```

Comment:
A locked section like this.

Reply:
Ah right! Yeah, I may have been a bit overzealous there. I'll try and remove some redundant locks.

Reply:
As far as I'm aware this is the only new lock I introduced. I removed it because you're right it isn't necessary. But I'm not sure what you meant with your "one per branch" remark earlier.

Reply:
No. That’s not what I meant. You definitely need a lock here. What I meant is that each branch also locks while flushing, and the incrementing of the snapshot used to be done within that locked section. Now we have three such sections, one for the explicit branch flush, one for the inferred and one for incrementing the snapshot.

Reply:
> You definitely need a lock here.

I should have looked better - I thought it could be removed because commitInternal already holds a lock. But that's not the case. I've reverted this change. 

> What I meant is that each branch also locks while flushing, and the incrementing of the snapshot used to be done within that locked section. Now we have three such sections, one for the explicit branch flush, one for the inferred and one for incrementing the snapshot.

I'm still not quite sure what problem you're seeing. There's a lock obtained in several places, yes. Why would that be a problem?

Reply:
I’m worried that it would potentially allow two transactions to flush at the same time to the same snapshot, then one gets incremented (making the changes visible) and if the other gets preempted between flushing the inferred and explicit, then half that transaction will be visible.

Reply:
Ah, I see what you mean now. Hm, yes. I need to think about this. It's possible we can do some magic with obtaining extra locks in the flush that we don't abandon until when increment is called. But it's getting a bit hairy, logic-wise, and moreover if we do that we effectively lose concurrent write access - _every_ transaction would block until the one that has the lock is committed or abandoned (which defeats the purpose of the whole SailSourceBranch model). 

_sigh_ I'll take a closer look, but we may need to abandon this approach after all. It's not been waste of time though, it gave me an opportunity to reacquaint myself with this part of the code.